### PR TITLE
WIP: Athul/redblacktree

### PR DESCRIPTION
**Pull request by athulappadan**
This is a simple implementation of red black tree. Looking forward for comments.

### CODE REVIEW

File: src/redBlackTree.jl [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+abstract Tree{K,V}
+
+type EmptyTree{K,V} <: Tree{K,V}
+end
+
+type RBNode{K,V} <: Tree{K,V}
+    key :: K
+    data :: V
+    left :: Tree{K,V}
+    right :: Tree{K,V}
```
Comment (kmsquire) [inline src/redBlackTree.jl]: This is probably the most straightforward implementation.  Performance-wise, however, in Julia, it's a [bad idea to have abstract members of types](http://docs.julialang.org/en/release-0.4/manual/performance-tips/#avoid-fields-with-abstract-type).

A workaround: https://github.com/JuliaLang/julia/issues/15791#issuecomment-207060497

(You'll probably have to read the whole thread for context.)

**Comment by kmsquire** [comment]
Seems like a reasonable start.  It would be good to

1. add some tests, to show that it works.  Probably also need to be able to display the tree somehow.
2. look at the tests for AVL trees, and compare the performance.

**Comment by athulappadan** [comment]
Thanks for the suggestions. Will commit the changes soon.

**Comment by athulappadan** [comment]
I have been going through the docs and different packages to understand how to efficiently define types, but still haven't got a grip on it. Main doubts are regarding abstract types, defining types recursively, parametric types, performance etc. Can anybody suggest some materials, papers to get some clarity on the julia type system..?

**Comment by timholy** [comment]
There are several good books and other learning resources, including those mentioned [here](http://julialang.org/learning/).

**Comment by kmsquire** [comment]
@athulappadan, I'm going to close this, but feel free to ping back here if you work on this some more, and someone can reopen it.  Or, you can just open another pull request.

**Pull request closed without merging**